node/node.go

```diff
@@ -130,6 +133,82 @@ func StateProvider(stateProvider statesync.StateProvider) Option {
 	}
 }
 
+// This function is expected to synchronize CometBFTs
+// stores with the application after statesync has been performed offline
+// It is expected that the blockstore and statestore are empty at
+// the time the function is called
+func BootstrapState(height uint64, appHash []byte, config *cfg.Config) error {
+	ctx := context.Background()
+	logger := log.NewTMLogger(log.NewSyncWriter(os.Stdout))
+	if config == nil {
+		config = cfg.DefaultConfig()
+	}
+	blockStore, stateDB, err := initDBs(config, cfg.DefaultDBProvider)
+
+	if err != nil {
+		return err
+	}
+
+	stateStore := sm.NewStore(stateDB, sm.StoreOptions{
+		DiscardABCIResponses: config.Storage.DiscardABCIResponses,
+	})
+
+	if !blockStore.IsEmpty() {
+		logger.Error("overwriting non empty block store")
+	}
+	genState, _, err := LoadStateFromDBOrGenesisDocProvider(stateDB, DefaultGenesisDocProviderFunc(config))
+	if err != nil {
+		return err
+	}
+
+	stateProvider, err := statesync.NewLightClientStateProvider(
+		ctx,
+		genState.ChainID, genState.Version, genState.InitialHeight,
+		config.StateSync.RPCServers, light.TrustOptions{
+			Period: config.StateSync.TrustPeriod,
+			Height: config.StateSync.TrustHeight,
+			Hash:   config.StateSync.TrustHashBytes(),
+		}, logger.With("module", "light"))
+	if err != nil {
+		return fmt.Errorf("failed to set up light client state provider: %w", err)
+	}
+
+	state, err := stateProvider.State(ctx, height)
+	if err != nil {
+		return err
+	}
+
+	commit, err := stateProvider.Commit(ctx, height)
+	if err != nil {
+		return err
+	}
+
+	if err := stateStore.Bootstrap(state); err != nil {
+		return err
+	}
+
+	// store.SaveBlockStoreState(&cmtstore.BlockStoreState{
```

Comment:
Does it work without this line?

Reply:
From checking the statesync code atm, I did not find a place where `base` and `height` are higher than 0 after statesync. I tested this locally in one of our e2e tests but am writing a test to commit as well. Even in the initial handshake if `store.LastBlockHeight == 0` we just check the `appHash` and do not return an error.  But I am writing a test to confirm this and to confirm that bootstrapping like this after statesync will work (I want to run statesync, crash the node, delete the stores and run this).

Reply:
But this is only correct if the store is empty beforehand (which is stated as expectation in the spec), and that is why Sergio is right above, that non empty stores should lead to an error indicating to the operator that they have to make sure this condition is met.

Reply:
@yihuang , you are in fact right, due to the way we boot up the Blocksync reactor [here](https://github.com/cometbft/cometbft/blob/70d0a65993c622adeb1cb6bfe5b48d9acd117458/blocksync/reactor.go#L72C15-L72C15), there will be an error if we do not call the code I commented out above. 

You actually said this in your other PR, that if a crash happens even in regular state-sync before the block is synced, the node will not start-up. I reproduced the case and now understand what you meant, apologies. But by definition this should not happen and we should have a more graceful handling of this. Setting the height and base like this violates their definition :"Height of first and last **block** in the block store" .

This is a bug and we should address it.